wip: fix various parts of code base to work with the latest version of blockly

p0lygun opened the pull request.

MagneticNeedle committed:
chore:update package-lock

MagneticNeedle committed:
feat: hook to `_Blockly.getMainWorkspace().showContextMenu` instead of `_Blockly.ContextMenu.show`;  LennardF1989/BF2042-Portal-Extensions#28

MagneticNeedle committed:
fix: use `_Blockly.getSelected` instead of `_Blockly.selected` as the latter is deprecated

MagneticNeedle committed:
fix: use mutation oberver to track if blockly is recreated and fire the init events; fixes LennardF1989/BF2042-Portal-Extensions#29

MagneticNeedle committed:
feat: for copy migrate to JSON serialization instead of XML

As XML has been ice boxed in blockly and as reported in LennardF1989/BF2042-Portal-Extensions#30, it fails to load team and player variables correctly, so migrating to json serialization is the obvious choice

MagneticNeedle committed:
feat: added ability to load serialized JSON

MagneticNeedle committed:
feat: export workspace as json

MagneticNeedle committed:
fix: use `serialization.workspaces.load`

MagneticNeedle committed:
feat: add spacing to copied/exported json

MagneticNeedle committed:
feat: add support for legacy  json import

MagneticNeedle committed:
feat: add support for pasting legacy xml

MagneticNeedle committed:
feat: dont add next block and coords when copying

MagneticNeedle committed:
fix: multiple blocks can now be selected

in v9 workspace.blocksDB_ is not blocksDB

The0zzy commented on src/web/app.js:
```diff
@@ -653,6 +657,25 @@ BF2042Portal.Extensions = (function () {
         }
 
+        const saveTemplateAsFile = (filename, dataObjToWrite) => {
+            const blob = new Blob([JSON.stringify(dataObjToWrite, null, 4)], { type: "text/json" });
```
Due to the json structure of blockly an indent of 2 or even only 1 would probably be better

p0lygun replied:
ok let's do 2 then

MagneticNeedle committed:
feat: paste blocks at mouse coords

MagneticNeedle committed:
fix: un-highlight only selected blocks

MagneticNeedle committed:
chore: when copying JSON only add one space

MagneticNeedle committed:
feat: only initialize BlocksLookup when its empty

MagneticNeedle committed:
feat: hook `_Blockly.inject` function

This is done so that when every the blockly is initialized again the plugins can also be re-initialized

MagneticNeedle committed:
fix: hook context menu when workspace is re-initialized

MagneticNeedle committed:
fix: multi-select  copy and paste now works properly
